modules/core/src/main/scala/parser.scala

```diff
@@ -312,71 +316,145 @@ object GraphQLParser {
     val initial    = ('A' to 'Z').toSet ++ ('a' to 'z').toSet + '_'
     val subsequent = initial ++ ('0' to '9').toSet
-    token((satisfy(initial), many(satisfy(subsequent))).mapN((h, t) => Ast.Name((h :: t).mkString)))
+    token(charIn(initial) ~ many(charIn(subsequent))).map {
+      case (h, t) => Ast.Name((h :: t).mkString)
+    }
   }
 }
 
 object CommentedText {
 
-  def skipWhitespace: Parser[Unit] =
-    takeWhile(c => c.isWhitespace || c == ',').void named "whitespace"
+  lazy val whitespace: Parser[Char] = charWhere(_.isWhitespace)
+
+  def opt[A](parser: Parser0[A]): Parser0[Option[A]] = parser.backtrack.?
+
+  def many[A](parser: Parser[A]): Parser0[List[A]] = parser.rep0
+
+  def skipWhitespace: Parser0[Unit] =
+    charsWhile0(c => c.isWhitespace || c == ',').void.withContext("whitespace")
 
   /** Parser that consumes a comment */
   def comment: Parser[Unit] = {
-    (char('#') ~> many(noneOf("\n\r")) <~ oneOf("\n\r") <~ skipWhitespace).void
+    (char('#') *> many((charWhere(c => c != '\n' && c != '\r'))) <* charIn('\n', '\r') <* skipWhitespace).void.withContext("comment")
   }
 
   /** Turns a parser into one that skips trailing whitespace and comments */
   def token[A](p: Parser[A]): Parser[A] =
-    p <~ skipWhitespace <~ many(comment)
+    p <* skipWhitespace <* many(comment)
+
+  def token0[A](p: Parser0[A]): Parser0[A] =
+    p <* skipWhitespace <* many(comment)
 
   /**
    * Consumes `left` and `right`, including the trailing and preceding whitespace,
    * respectively, and returns the value of `p`.
    */
-  private def _bracket[A,B,C](left: Parser[B], p: => Parser[A], right: => Parser[C]): Parser[A] =
-    token(left) ~> token(p) <~ token(right)
+  private def _bracket[A,B,C](left: Parser[B], p: => Parser0[A], right: => Parser[C]): Parser[A] =
+    token(left) *> token0(p) <* token(right)
 
   /** Turns a parser into one that consumes surrounding parentheses `()` */
-  def parens[A](p: => Parser[A]): Parser[A] =
-    _bracket(char('('), p, char(')')).named(s"parens(${p.toString})")
+  def parens[A](p: => Parser0[A]): Parser[A] =
+    _bracket(char('('), p, char(')')).withContext(s"parens(${p.toString})")
 
   /** Turns a parser into one that consumes surrounding curly braces `{}` */
-  def braces[A](p: => Parser[A]): Parser[A] = {
+  def braces[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('{'), q, char('}')).named(s"braces(${q.toString})")
+    _bracket(char('{'), q, char('}')).withContext(s"braces(${q.toString})")
   }
 
   /** Turns a parser into one that consumes surrounding square brackets `[]` */
-  def squareBrackets[A](p: => Parser[A]): Parser[A] = {
+  def squareBrackets[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('['), q, char(']')).named(s"squareBrackets(${q.toString})")
+    _bracket(char('['), q, char(']')).withContext(s"squareBrackets(${q.toString})")
+  }
+}
+
+object Literals {
+
+  lazy val stringLiteral: Parser[String] = {
```

Comment:
I think you should not default to `lazy val` and should instead only use it when you really need it. Looks like everything is getting lazy val here.